ceph/ceph.go

```diff
@@ -178,123 +256,158 @@ func matchSlice(a, b []string) bool {
 	}
 
+	outMetrics := []plugin.Metric{}
 	for _, daemon := range ceph.daemons {
+		// Retrieve daemon metric values
 		if dmetric, err := ceph.GetCephDaemonMetrics(mts, daemon); err == nil {
-			metrics = append(metrics, dmetric...)
+			outMetrics = append(outMetrics, dmetric...)
 		}
 	}
 
-	return metrics, nil
+	return outMetrics, nil
 }
 
 // GetMetricTypes returns the metric types exposed by ceph-daemon sockets
-func (ceph *Ceph) GetMetricTypes(cfg plugin.ConfigType) ([]plugin.MetricType, error) {
-	mts := []plugin.MetricType{}
-	for _, metricMeta := range allMetrics {
-		mts = append(mts, plugin.MetricType{Namespace_: createNamespace(metricMeta.ns), Description_: metricMeta.description})
+func (ceph *Ceph) GetMetricTypes(cfg plugin.Config) ([]plugin.Metric, error) {
+	mts := []plugin.Metric{}
+
+	// init ceph plugin with Config settings (only once)
+	if err := ceph.Init(cfg); err != nil {
+		return nil, err
+	}
+
+	// Get metric types for each daemon (if more daemon IDs, only first found ID will be scanned for available metrics)
+	daemonsCollected := make(map[string]bool)
+
+	for _, daemon := range ceph.daemons {
+		if !daemonsCollected[daemon.name] {
+			if dmetric, err := ceph.GetCephDaemonMetrics(nil, daemon); err == nil {
+				mts = append(mts, dmetric...)
+				daemonsCollected[daemon.name] = true
+			}
+		}
 	}
 
 	return mts, nil
 }
 
 // GetConfigPolicy returns a ConfigPolicy
-func (ceph *Ceph) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
-	c := cpolicy.New()
-	socketPath, _ := cpolicy.NewStringRule("socket_path", false, socketPathDefault)
-	socketPrefix, _ := cpolicy.NewStringRule("socket_prefix", false, socketPrefixDefault)
-	socketExt, _ := cpolicy.NewStringRule("socket_ext", false, socketExtDefault)
-	path, _ := cpolicy.NewStringRule("path", false, cephBinPathDefault)
-	p := cpolicy.NewPolicyNode()
-	p.Add(socketPath)
-	p.Add(socketPrefix)
-	p.Add(socketExt)
-	p.Add(path)
-	c.Add([]string{"intel", "storage", "ceph"}, p)
-	return c, nil
+func (ceph *Ceph) GetConfigPolicy() (plugin.ConfigPolicy, error) {
+	policy := plugin.NewConfigPolicy()
+	policy.AddNewStringRule([]string{"intel", "storage", "ceph"}, "socket_path", false, plugin.SetDefaultString(socketPathDefault))
+	policy.AddNewStringRule([]string{"intel", "storage", "ceph"}, "socket_prefix", false, plugin.SetDefaultString(socketPrefixDefault))
+	policy.AddNewStringRule([]string{"intel", "storage", "ceph"}, "socket_ext", false, plugin.SetDefaultString(socketExtDefault))
+	policy.AddNewStringRule([]string{"intel", "storage", "ceph"}, "path", false, plugin.SetDefaultString(cephBinPathDefault))
+	return *policy, nil
 }
 
 // getCephBinaryPath returns path to ceph executable
-func getCephBinaryPath(config map[string]ctypes.ConfigValue) string {
+func getCephBinaryPath(config plugin.Config) string {
 	// check global config as a first
-	if path, ok := config["path"]; ok {
-		return path.(ctypes.ConfigValueStr).Value
+	if path, err := config.GetString("path"); err == nil {
```

Comment:
`err` is not handled, 
There will be a problem if user creates config variable with inappropriate data type. It will be helpful to handle this error. Please see  other usage of `config.GetString()`